Introduce an identifier for the default source/target for a collection

isuruf opened the pull request:
And rename DEFAULT_SOURCE -> UNNAMED_SOURCE

alexfikl commented:
This looks good to me, but could use a bit more documentation. Some things that come to mind:
* Why can't we use `None` for `UNNAMED_SOURCE` and `UNNAMED_TARGET`?
* `DEFAULT_SOURCE` and `DEFAULT_TARGET` are replaced in `LocationTagger` with whatever it was given in the constructor, so not necessarily the `auto_source` and `auto_target` of the `GeometryCollection`. This disagrees with their docs?
* Are `DEFAULT_SOURCE` and `DEFAULT_TARGET` supposed to be for internal use now? i.e. stuff in `pytential.symbolic`.

isuruf commented:
> Why can't we use None for UNNAMED_SOURCE and UNNAMED_TARGET?

They need to be distinct values in order to be placed in a dictionary.

> DEFAULT_SOURCE and DEFAULT_TARGET are replaced in LocationTagger with whatever it was given in the constructor, so not necessarily the auto_source and auto_target of the GeometryCollection. This disagrees with their docs?

Fixed now.

> Are DEFAULT_SOURCE and DEFAULT_TARGET supposed to be for internal use now? i.e. stuff in pytential.symbolic.

Nope. UNNAMED_SOURCE and UNNAMED_TARGET are internal use only. Added to docs.

inducer commented on pytential/collection.py:
```diff
@@ -142,8 +142,8 @@ def __init__(self,
             By default, they are set to
-            :class:`~pytential.symbolic.dof_desc.DEFAULT_SOURCE` and
-            :class:`~pytential.symbolic.dof_desc.DEFAULT_TARGET` for
+            :class:`~pytential.symbolic.dof_desc.UNNAMED_SOURCE` and
+            :class:`~pytential.symbolic.dof_desc.UNNAMED_TARGET` for
```
...unspecified values...

inducer commented on pytential/symbolic/dof_desc.py:
```diff
@@ -23,6 +23,9 @@
 from typing import Any, Hashable, Optional, Union
 
 __doc__ = """
+.. autoclass:: UNNAMED_SOURCE
+.. autoclass:: UNNAMED_TARGET
+
```
```suggestion
```

inducer commented on pytential/symbolic/dof_desc.py:
```diff
@@ -55,12 +58,26 @@
 
 # {{{ discretizations
 
+class UNNAMED_SOURCE:                   # noqa: N801
```
Link to the use case in the interface for `prepare_...`. (same for `_UNNAMED_TARGET`)

inducer reviewed (approved):
LGTM subject to the changes above.

inducer commented on pytential/symbolic/execution.py:
```diff
@@ -37,6 +37,7 @@
 from pytential import sym
+from pytential.symbolic.dof_desc import UNNAMED_SOURCE, UNNAMED_TARGET
 
 import logging
```
- Delete `domains` arg on `build_matrix`
- Delete `domains` arg on `scipy_op`.

alexfikl replied:
Why? Weren't those supposed to support multiple geometries with different properties? (which, yeah, nobody uses or tests at all.. but still!)

inducer replied:
Hmmm, upon closer reflection, I think I may have jumped the gun on removing `domains`. I think it has a role to play still. What tripped me up is that there should be `source_domains` and `target_domains`, technically. Where it says `DEFAULT_TARGET` must be a key, instead say something like "target" must be the result of an unnamed binding, such as might result from passing a tuple to `bind`.

IOW, @alexfikl, you're completely right.

inducer commented:
Thx!
